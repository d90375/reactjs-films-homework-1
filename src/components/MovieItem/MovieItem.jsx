import React, { Component } from 'react';
import PropTypes from 'prop-types';
import WatchNowWindow from '../WatchNowWindow';
import ViewInfoWindow from '../ViewInfoWindow';

import styles from './MovieItem.scss';

class MovieItem extends Component {
  state = {
    isInfoOpen: false,
  };

  switchViewInfo = () => {
    this.setState(({ isInfoOpen }) => ({
      isInfoOpen: !isInfoOpen,
    }));
  }


  render() {
    const {
      film, fetchTrailer, setMoviesCondition,
    } = this.props;
    const { isInfoOpen } = this.state;

    const {
      id, poster, title, genres, rating,
    } = film;

    return (
      <div className={styles.container}>
        <div className={styles.preview}>
          <img className={styles.poster} src={poster} alt="poster" />
          <a
            className={styles.link}
            href={`/details/${id}`}
            name="link"
            onClick={() => {
              setMoviesCondition('Trending');
            }}
          >
            <div className={styles.info}>
              <h3 className={styles.title}>
                {title.length < 15 ? title : `${title.substring(0, 13)}...`}
              </h3>
              <p className={styles.rating}>{rating}</p>
              <p className={styles.genres}>{genres}</p>
            </div>
          </a>
        </div>
        <WatchNowWindow
          name="watch"
          film={film}
          switchViewInfo={this.switchViewInfo}
          fetchTrailer={fetchTrailer}
        />
        { isInfoOpen
          ? (
            <ViewInfoWindow
              film={film}
              switchViewInfo={this.switchViewInfo}
              fetchTrailer={fetchTrailer}
            />
          )
          : null }
      </div>
    );
  }
}

MovieItem.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number,
    poster: PropTypes.string,
    title: PropTypes.string,
    genres: PropTypes.string,
    rating: PropTypes.number,
  }).isRequired,
  fetchTrailer: PropTypes.func.isRequired,
  setMoviesCondition: PropTypes.func.isRequired,
};

export default MovieItem;
