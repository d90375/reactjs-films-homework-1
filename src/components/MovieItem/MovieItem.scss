.container {
  position: relative;
  width: 250px;
  height: 320px;
  margin: 0 10px 20px 10px;
  border-radius: 6px;
  filter: drop-shadow(0 10px 20px rgba(0, 0, 0, 0.04));
  background-color: #b3b3b3;
  overflow: hidden;

  @media (min-width: 390px) and (max-width: 659px) {
    width: 100%;
    height: auto;
    border-radius: 10px;
  }
}

.preview {
  &:hover ~ div {
    display: flex;
  }
}

.poster {
  width: 100%;
  border-radius: 6px;
  overflow: hidden;
}

.link {
  cursor: pointer;
}

.info {
  position: absolute;
  bottom: 0;
  box-sizing: border-box;
  width: 250px;
  height: 100px;
  padding: 0 20px;
  background-color: white;

  @media (min-width: 390px) and (max-width: 659px) {
    width: 100%;
  }

  .title {
    display: inline-block;
    width: 150px;
    height: 22px;
    font-size: 16px;
    line-height: 22px;
    font-weight: bold;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    text-transform: uppercase;
    color: #242b2d;

    @media (min-width: 390px) and (max-width: 659px) {
      width: calc(100% - 40px);
    }
  }

  .rating {
    display: inline-block;
    width: 30px;
    font-size: 12px;
    line-height: 20px;
    font-weight: bold;
    vertical-align:top;
    color: #47c3f0;
    text-align: center;
    border-radius: 6px;
    border: 2px solid #49c1f0;
  }

  .genres {
    margin: 0;
    font-size: 12px;
    line-height: 16px;
    letter-spacing: 1px;
    color: #47c3f0;
    font-weight: bold;
  }
}
