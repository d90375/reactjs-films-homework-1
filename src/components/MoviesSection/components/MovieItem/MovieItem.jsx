import React from 'react';
import PropTypes from 'prop-types';
import styles from './MovieItem.scss';

const MovieItem = ({
  film: {
    poster, title, genres, rating,
  },
}) => (
  <div className={styles.container}>
    <img className={styles.poster} src={poster} alt="poster" />
    <div className={styles.info}>
      <h3 className={styles.title}>
        {title.length < 18 ? title : `${title.substring(0, 15)}...`}
      </h3>
      <p className={styles.rating}>{rating}</p>
      <p className={styles.genres}>{genres}</p>
    </div>
  </div>
);

MovieItem.propTypes = {
  film: PropTypes.shape({
    poster: PropTypes.string,
    title: PropTypes.string,
    genres: PropTypes.string,
    rating: PropTypes.number,
  }).isRequired,
};

export default MovieItem;
