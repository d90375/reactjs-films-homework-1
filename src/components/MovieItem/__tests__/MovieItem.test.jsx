import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';
import MovieItem from '../MovieItem';

describe('MovieItem tests', () => {
  const film = {
    title: 'title',
    genres: 'action, crime',
    rating: '8',
    poster: 'url',
  };

  describe('MovieItem render', () => {
    it('MovieItem renders correctly', () => {
      const mockCallBack = jest.fn();
      const renderer = new ShallowRenderer();
      renderer.render(<MovieItem film={film} fetchTrailer={mockCallBack} />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });

    it('MovieItem renders correctly', () => {
      const mockCallBack = jest.fn();
      film.title = 'title!!!!!!!!!!!!!!!!!!!!!!!!!!!!!';
      const tree = create(
        <MovieItem film={film} fetchTrailer={mockCallBack} />,
      );

      const watchNowWindow = tree.root.findByProps({ name: 'watch' });
      watchNowWindow.props.switchViewInfo();
      expect(tree).toMatchSnapshot();
    });
  });
});
