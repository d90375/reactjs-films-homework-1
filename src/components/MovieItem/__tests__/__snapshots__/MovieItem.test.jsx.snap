// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem tests MovieItem render MovieItem renders correctly with movie title length less then 15 characters 1`] = `
<div
  className="container"
>
  <div
    className="preview"
  >
    <img
      alt="poster"
      className="poster"
      src="url"
    />
    <Link
      name="link"
      onClick={[Function]}
      to="/details/undefined"
    >
      <div
        className="info"
      >
        <h3
          className="title"
        >
          title
        </h3>
        <p
          className="rating"
        >
          8
        </p>
        <p
          className="genres"
        >
          action, crime
        </p>
      </div>
    </Link>
  </div>
  <WatchNowWindow
    fetchTrailer={[MockFunction]}
    film={
      Object {
        "genres": "action, crime",
        "poster": "url",
        "rating": 8,
        "title": "title",
      }
    }
    name="watch"
    switchViewInfo={[Function]}
  />
</div>
`;

exports[`MovieItem tests MovieItem render MovieItem renders correctly with movie title length more then 15 characters 1`] = `
<div
  className="container"
>
  <div
    className="preview"
  >
    <img
      alt="poster"
      className="poster"
      src="url"
    />
    <a
      href="/details/undefined"
      name="link"
      onClick={[Function]}
    >
      <div
        className="info"
      >
        <h3
          className="title"
        >
          title!!!!!!!!...
        </h3>
        <p
          className="rating"
        >
          8
        </p>
        <p
          className="genres"
        >
          action, crime
        </p>
      </div>
    </a>
  </div>
  <div
    className="container"
  >
    <button
      className="watch"
      onClick={[Function]}
      type="button"
    >
      ►
    </button>
    <p>
      Watch Now
    </p>
    <button
      className="button secondary"
      onClick={[Function]}
      type="button"
    >
      View Info
    </button>
  </div>
  <div
    className="container"
  >
    <img
      alt="poster"
      className="poster"
      src="url"
    />
    <div
      className="wrapper"
    >
      <button
        className="close"
        onClick={[Function]}
        type="button"
      >
        ×
      </button>
      <div
        className="info"
      >
        <h3
          className="title"
        >
          title!!!!!!!!!!...
        </h3>
        <p
          className="rating"
        >
          8
        </p>
        <p
          className="genres"
        >
          action, crime
        </p>
      </div>
      <p
        className="overview"
      />
      <button
        className="button primary"
        onClick={[Function]}
        type="button"
      >
        Watch Now
      </button>
    </div>
  </div>
</div>
`;
